Migration to LLVM 15

Focshole opened the pull request:
Integrated marcof's JIT test, updated imports and migrated to ORC v2. Migration to LLVM 13 with compatibility up to LLVM 15

skeru commented on CMakeLists.txt:
```diff
@@ -108,9 +108,9 @@ if (NOT CMAKE_BUILD_TYPE)
-set ( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -fno-rtti -fvisibility-inlines-hidden" )
-set ( CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DVC_DEBUG -fvisibility-inlines-hidden" )
-set ( CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -fPIC -fvisibility-inlines-hidden" )
+set ( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -ldl -fno-use-cxa-atexit" )
```
why linking flags here?

Focshole replied:
Probably they were not required. I'd remove them if you agree with me.

skeru replied:
I remember `-fvisibility-inline-hidden` was added to resolve potential conflicts with global variable names when using libVC as external library. Not sure if that is still required again. Did you investigate before removing it?

Focshole replied:
No, sorry, I didn't know the background which led to those edits, I just kept them untouched from marcof's fork

skeru replied:
might be worth setting [this variable](https://cmake.org/cmake/help/latest/variable/CMAKE_VISIBILITY_INLINES_HIDDEN.html#variable:CMAKE_VISIBILITY_INLINES_HIDDEN) to avoid exporting symbols that might create conflicts in the code using our library.

skeru replied:
About the other flag, my google-fu says that it might not be very useful. [Link](https://github.com/offa/stm32-eth/issues/23) In our case can you double check if we actually have a difference in the library binary (using gcc and using clang) with/without this flag?

Focshole replied:
> About the other flag, my google-fu says that it might not be very useful. [Link](https://github.com/offa/stm32-eth/issues/23) In our case can you double check if we actually have a difference in the library binary (using gcc and using clang) with/without this flag?

The binary files differ, I'd set the cmake flag then

Focshole replied:
> About the other flag, my google-fu says that it might not be very useful. [Link](https://github.com/offa/stm32-eth/issues/23) In our case can you double check if we actually have a difference in the library binary (using gcc and using clang) with/without this flag?

Both testing binaries and the library binary show a few differences, so I'd guess `-fno-use-cxa-atexit` is needed.

Focshole replied:
> I remember `-fvisibility-inline-hidden` was added to resolve potential conflicts with global variable names when using libVC as external library. Not sure if that is still required again. Did you investigate before removing it?

https://cmake.org/cmake/help/latest/policy/CMP0063.html
I'd guess it is ok to remove it but I'm not 100% sure

joaobispo replied:
I would suggest, if you discover the reason, to add that information in the code as a comment

skeru replied:
> > I remember `-fvisibility-inline-hidden` was added to resolve potential conflicts with global variable names when using libVC as external library. Not sure if that is still required again. Did you investigate before removing it?
> 
> https://cmake.org/cmake/help/latest/policy/CMP0063.html I'd guess it is ok to remove it but I'm not 100% sure

We can remove this explicit flag if we set the corresponding property in the cmake, and set cmake_minimum_version to as high as this policy change.

Focshole replied:
> I would suggest, if you discover the reason, to add that information in the code as a comment

Found it, https://gcc.gnu.org/wiki/Visibility#line-14 , closing this

skeru commented on README.md:
```diff
@@ -11,15 +11,15 @@ An introductory video is available on [YouTube](https://www.youtube.com/watch?v=
 ## Dependencies
 libVersioningCompiler requires:
- - Ubuntu 16.04 or greater
+ - Ubuntu 20 or greater
```
20.04 or 20.10?

Focshole replied:
20.04 LTS, I'll fix it

skeru commented on cmake/FindLLVM.cmake:
```diff
@@ -8,20 +8,41 @@
-  3.3
-  3.2
-  3.1)
+set (LIBCLANG_KNOWN_LLVM_VERSIONS 15.0.0
```
should be replaced with a better findLLVM

skeru replied:
you should be able to use the `find_package( ... REQUIRED )` (in the main cmake file) to harvest the information about installed LLVM versions. This `cmake/FindLLVM` should no longer be required.
Backward compatibility up to LLVM 3.1 looks like something we don't have anyway in the rest of the code.

Which exact components are mapped into the old libraries we were using a decade ago... is certainly a good question.

Focshole replied:
It is mainly used to find `llvm-config` and to get the proper recommend flags. The name is a bit misleading imho.

I remove older versions because they are useless here, I agree with you. Probably there is something to do in `FindLibClang.cmake` rather than on this. Feel free to reopen the issue if I am mistaken/ I have missed something.

skeru commented on cmake/FindLLVM.cmake:
```diff
@@ -32,11 +53,13 @@ set (llvm_lib_search_paths
 
 foreach (version ${LIBCLANG_KNOWN_LLVM_VERSIONS})
   string(REPLACE "." "" undotted_version "${version}")
+  string(REPLACE ".0.0" "" stripped_version "${version}")
```
does not look right. What case is it handling?

Focshole replied:
With versions like LLVM 13.0.0, it used to look for libraries with llvm1300 or llvm13.0.0 in their path, without considering llvm13

skeru commented on cmake/FindLibClang.cmake:
```diff
@@ -85,36 +117,70 @@ find_library (LIBCLANG_LIBRARY NAMES libclang.imp libclang clang
 
 MACRO(FIND_AND_ADD_CLANG_LIB _libname_)
-find_library(CLANG_${_libname_}_LIB
+find_library(CLANG_${_libname_}_LIB 
```
no whitespace please

skeru commented on include/versioningCompiler/Compiler.hpp:
```diff
@@ -97,25 +97,25 @@ class Compiler
+                                 void ** handler);
 
-/** \brief Closes the binary shared object, set *handler to nullptr, and
+    virtual /** \brief Closes the binary shared object, set *handler to nullptr, and
```
```suggestion
    virtual
    /** \brief Closes the binary shared object, set *handler to nullptr, and
```

skeru commented on include/versioningCompiler/Compiler.hpp:
```diff
@@ -97,25 +97,25 @@ class Compiler
  * Output of this method is supposed to reinterpreted by the caller.
  */
-  std::vector<void*> loadSymbols(const std::string &bin,
+ virtual std::vector<void*> loadSymbols(std::string &bin,
```
why removing `const` here?

Focshole replied:
`std::string &bin` was being used as a handler for the JITCompiler implementation. I have just changed JITCompiler implementation in order to keep this as const.

skeru commented on include/versioningCompiler/CompilerImpl/ClangLLVM/OptUtils.hpp:
```diff
@@ -263,7 +268,8 @@ static void AddOptimizationPasses(llvm::legacy::PassManagerBase &MPM,
   }
-  Builder.DisableUnitAtATime = !UnitAtATime;
+  // NOTE: The DisableUnitAtATime switch has been removed. https://llvm.org/doxygen/PassManagerBuilder_8cpp_source.html#l00828 
+  //Builder.DisableUnitAtATime = !UnitAtATime;
```
just remove the code

skeru commented on include/versioningCompiler/CompilerImpl/ClangLLVM/OptUtils.hpp:
```diff
@@ -282,9 +288,11 @@ static void AddOptimizationPasses(llvm::legacy::PassManagerBase &MPM,
     TM->adjustPassManager(Builder);
   }
+  #if LLVM_MAJOR_VERSION <15
   if (Coroutines) {
-    llvm::addCoroutinePassesToExtensionPoints(Builder);
+    llvm::addCoroutinePassesToExtensionPoints(Builder); // todo No clue on how to replace this for LLVM-15
   }
+  #endif
```
this was a sort of clone from mainstream `opt`. What do they do there?

Focshole replied:
I'd guess it was a sort of a clone from [this](https://github.com/llvm/llvm-project/blob/release/14.x/llvm/tools/opt/opt.cpp), right?
This changes quite a lot between  [14](https://github.com/llvm/llvm-project/blob/release/14.x/llvm/tools/opt/opt.cpp) and  [15](https://github.com/llvm/llvm-project/blob/release/15.x/llvm/tools/opt/opt.cpp). I guess it would not be retrocompatible anyway.

skeru replied:
Well, yes. Ten versions earlier, but yes.

Retrocompatibility of this class is probably hard to obtain.
Does the 14 version work well when compiled against LLVM 15?

You could test it by scheduling a generic optimisation pass, say `mem2reg` in a test case.
If that does not work, I suggest to selectively disable this Compiler Implementation from cmake, depending on the version of LLVM.

Focshole replied:
Well, [Test.cpp](https://github.com/skeru/libVersioningCompiler/pull/3/files#diff-bce024b33fc3cc08c6ad47c2fca5f2a12154e381a0c4345c94dc611c9543721dR115) seems to schedule an optimization pass and both tests (libVC_test and libVC_testJit) succeed on llvm 15. Can it be considered as working?

skeru replied:
I would call it a fair test.

However, can you please investigate was was saying the commit which removed these lines in mainstream LLVM?
I suspect that it was integrated into something new, but just to be sure...

Focshole replied:
That's a good point, I'll look into that!

Focshole replied:
Updated! It is properly working for LLVM 13, 14 and 15!

skeru commented on include/versioningCompiler/CompilerImpl/JITCompiler.hpp:
```diff
@@ -0,0 +1,153 @@
+#include "llvm/ExecutionEngine/SectionMemoryManager.h"
+#include "llvm/ExecutionEngine/Orc/CompileUtils.h"
+#include "llvm/ExecutionEngine/Orc/IRCompileLayer.h"
+//#include "llvm/ExecutionEngine/Orc/ExecutorProcessControl.h"
```
```suggestion
```

skeru commented on CMakeLists.txt:
```diff
@@ -173,7 +176,7 @@ set (VC_LIB_HDR ${VC_LIB_HDR1} ${VC_LIB_HDR2} ${VC_LIB_HDR3} )
 
 add_library( ${VC_LIB_NAME} STATIC ${VC_LIB_SRC} ${VC_LIB_HDR} )
-target_link_libraries( ${VC_LIB_NAME} ${VC_LIB_DEPS} )
+target_link_libraries( ${VC_LIB_NAME} ${VC_LIB_DEPS} z pthread tinfo)
```
can we make a self_explanatory variable for these new additions?

Focshole replied:
Probably they were not required. I'd remove them if you agree with me.

skeru commented on CMakeLists.txt:
```diff
@@ -108,9 +108,9 @@ if (NOT CMAKE_BUILD_TYPE)
   set ( CMAKE_BUILD_TYPE Release )
 endif()
 
-set ( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -fno-rtti -fvisibility-inlines-hidden" )
```
we should set a CXX standard elsewhere, possibly with modern cmake commands

Focshole replied:
Do you mean something like [this](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html)?

skeru replied:
More like [`CMAKE_CXX_STANDARD`](https://cmake.org/cmake/help/latest/variable/CMAKE_CXX_STANDARD.html#variable:CMAKE_CXX_STANDARD)

Focshole replied:
Should be solved

skeru reviewed (commented):
Thank you for the great work.
Some comments as of now.
Will later test and provide more insights.

skeru commented on README.md:
```diff
@@ -28,15 +28,28 @@ like the Clang-as-a-library compiler implementation.
+chmod +x llvm.sh
+[sudo] ./llvm.sh ${LLVM_V} all
+[sudo] apt-get install -y git cmake uuid-dev g++-10 libz-dev
+[optional][sudo] update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-15 100
+[optional][sudo] update-alternatives --install /usr/bin/clang clang /usr/bin/clang-15 100
+[optional][sudo] update-alternatives --install /usr/bin/opt opt /usr/bin/opt-15 100
```
IIRC these can be manually set in the source code.

In the long run, the configure system (cmake) can automatically detect them and set some default values.
As of now, however, these alias are not really required neither beneficial. Unless you also changed something that relies on them elsewhere.

Focshole replied:
Without those on Ubuntu 20.04 LTS I had some internal errors (in libc-main and in lto/lto-partition.c) in the linking process which were practically undebuggable. I suspected they happened due to incoherent versions of clang with opt but I couldn't find the culprit. IMHO I have poorly set up Ubuntu, resulting in a broken configuration, that's why I had added those commands.

skeru replied:
If you can reproduce the error in a clean environment, we can have a call and try to solve the issue together. This solution may work, but is sounds not the appropriate way to handle it.

skeru commented on README.md:
```diff
@@ -28,15 +28,28 @@ like the Clang-as-a-library compiler implementation.
 mkdir build
 cd build
-cmake .. [-DCMAKE_INSTALL_PREFIX="/path/to/your/custom/install/folder/"]
+CXX=/bin/g++-10 cmake .. [-DCMAKE_INSTALL_PREFIX="/path/to/your/custom/install/folder/"]
```
Not sure why you need to manually specify the compiler.

If you need so, I would prefer to move it as cmake variable rather than env variable. I.e. `-DCMAKE_CXX_COMPILER=...`

Focshole replied:
Same as [here](https://github.com/skeru/libVersioningCompiler/pull/3?notification_referrer_id=NT_kwDOAh-ycbM0NTExNTQ2NTY4OjM1NjMxNzI5#discussion_r985247312)

skeru replied:
by default, cmake should try to use the command `cxx` to compile cpp files, and `cxx` should always point to a valid cpp compiler installed in the system. The readme should not suggest the user to redirect the compilation to any specific compiler unless the suggested one is the only compatible compiler due to intrinsics in the code (and I think this is not our case)

skeru commented on CMakeLists.txt:
```diff
@@ -107,10 +107,10 @@ if (NOT CMAKE_BUILD_TYPE)
-set ( CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DVC_DEBUG -fvisibility-inlines-hidden" )
-set ( CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -fPIC -fvisibility-inlines-hidden" )
+set ( CMAKE_CXX_STANDARD 14 )
+set ( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -ldl -fno-use-cxa-atexit" )
```
The linking of `dl` library should be specified in `target_link_libraries`

Focshole replied:
It is actually specified in `target_link_libraries`, I guess it is redundant (diff between the two executables shows that they're the same), I'll remove it

skeru commented on CMakeLists.txt:
```diff
@@ -107,10 +107,10 @@ if (NOT CMAKE_BUILD_TYPE)
+set ( CMAKE_CXX_STANDARD 14 )
+set ( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -ldl -fno-use-cxa-atexit" )
+set ( CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DVC_DEBUG -fno-rtti -fno-use-cxa-atexit" )
+set ( CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -fpic -fno-rtti -fno-use-cxa-atexit" )
```
broadly speaking, cmake best practices now suggests that touching directly the `CMAKE_YYY_FLAGS_YYY` is not a good idea, and possibly there are equivalent, more specific options to automatically update them, like setting a target property.

Can you please investigate? If this is requiring more than a couple hours effort, please ignore and resolve.

skeru commented on TestJit.cpp:
```diff
@@ -0,0 +1,176 @@
+#include <string>
+#include <vector>
+#include <iostream>
+#include <stdlib.h>
```
Can you please sort them alphabetically?

skeru commented on TestJit.cpp:
```diff
@@ -0,0 +1,176 @@
+#if HAVE_CLANG_AS_LIB
+
+#include "versioningCompiler/CompilerImpl/ClangLibCompiler.hpp"
+#include <versioningCompiler/CompilerImpl/JITCompiler.hpp>
```
why `" "` and `< >` ?

skeru commented on lib/CompilerImpl/SystemCompiler.cpp:
```diff
@@ -62,10 +62,10 @@ bool SystemCompiler::hasOptimizer() const
 // ------------------------------- generate IR --------------------------------
 // ----------------------------------------------------------------------------
-std::string SystemCompiler::generateIR(const std::vector<std::string> &src,
+  std::string SystemCompiler::generateIR(const std::vector<std::string> &src,
```
?

Focshole commented on CMakeLists.txt:
```diff
@@ -107,10 +108,15 @@ if (NOT CMAKE_BUILD_TYPE)
+set ( CMAKE_VISIBILITY_INLINES_HIDDEN ON )
+set ( CMAKE_POSITION_INDEPENDENT_CODE ON )
+set ( CMAKE_INTERPROCEDURAL_OPTIMIZATION ON)  
+set ( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-use-cxa-atexit -fno-rtti" )
```
https://gitlab.kitware.com/cmake/cmake/-/issues/20610 -fno-rtti it is still required.

skeru commented on .travis.yml:
```diff
@@ -0,0 +1,62 @@
+          sources:
+            - llvm-toolchain-focal-15
+            - sourceline: "deb http://apt.llvm.org/focal/ llvm-toolchain-focal-15 main"
+              key_url: "https://apt.llvm.org/llvm-snapshot.gpg.key"
+          packages:
+            - cmake
+            - gcc-10
+            - g++-10
+            - libstdc++-10-dev
+            - libz-dev
+            - uuid-dev
+            - clang-15
+            - lldb-15
+            - lld-15
+            - clangd-15
+            - clang-format-15
+            - clang-tools-15
+            - llvm-15-dev
+            - lld-15
+            - lldb-15
+            - llvm-15-tools
+            - libomp-15-dev
+            - libc++-15-dev
+            - libc++abi-15-dev
+            - libclang-common-15-dev
+            - libclang-15-dev
+            - libclang-cpp15-dev
+            - libunwind-15-dev
+      env:
+        - ENV_SETUP="CC=clang-15 && CXX=clang-15"
+        - LLVM_CONFIG="llvm-config-15"
+        - BEFORE_INSTALL="sudo update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-15 100 && sudo update-alternatives --install /usr/bin/clang clang /usr/bin/clang-15 100 && sudo update-alternatives --install /usr/bin/opt opt /usr/bin/opt-15 100"
+        - RUN_TESTS="sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-10 100 && sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-10 100 && chmod +x libVC_test libVC_testJit libVC_testUtils && ./libVC_testUtils && ./libVC_test && ./libVC_testJit "
+
+before_install:
+  - eval "${ENV_SETUP}" && eval "${BEFORE_INSTALL}"
+
+
+script:
+  - mkdir build && cd build && cmake .. && make 
```
uhm . . .
Is this a side effect of the [git fire](https://juststickers.in/wp-content/uploads/2017/04/in-case-of-fire.png) procedure?

Focshole replied:
No no, it is for Travis CI https://app.travis-ci.com/github/Focshole/libVersioningCompiler/builds/256694521

skeru replied:
woops

skeru commented on CMakeLists.txt:
```diff
@@ -107,10 +108,16 @@ if (NOT CMAKE_BUILD_TYPE)
+set ( CMAKE_CXX_VISIBILITY_INLINES_HIDDEN ON )
+set ( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-use-cxa-atexit -fno-rtti" )
+set ( CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DVC_DEBUG" )
+set ( CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3" )
```
This line should be not required

Focshole replied:
I had changed to include it only if gcc/g++ is being used. In case it is wrong, open a review on the newer edit. Also, there is not a clear policy on how to set optimization flags so far https://gitlab.kitware.com/cmake/cmake/-/issues/23374

skeru commented on CMakeLists.txt:
```diff
@@ -107,10 +108,16 @@ if (NOT CMAKE_BUILD_TYPE)
+set ( CMAKE_INTERPROCEDURAL_OPTIMIZATION ON)  
+set ( CMAKE_CXX_VISIBILITY_INLINES_HIDDEN ON )
+set ( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-use-cxa-atexit -fno-rtti" )
+set ( CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DVC_DEBUG" )
```
This one can be moved to source_file_property > compile flag, under conditional "Debug" CMAKE_BUILD_TYPE

skeru commented on Test.cpp:
```diff
@@ -103,6 +103,7 @@ int main(int argc, char const *argv[])
   // ---------- End compilers initialization ----------
   // start configuring version v
   builder._compiler = cc;
+  builder._genIROptionList={vc::Option("fpic", "-fPIC")};
```
is this really required?

Focshole replied:
I think so, otherwise i get the following result while running libVC_test:
```
Notify: v3 IR prepared. Going for v3 compilation
Error: compilation 3 failed.
```
with error:
```
/usr/bin/ld: /tmp/opt_IR_35...a.o: warning: relocation against `global_var' in read-only section `.text'
/usr/bin/ld: /tmp/opt_IR_35...a.o: relocation R_X86_64_PC32 against symbol `global_var' can not be used when making a shared object; recompile with -fPIC
```

skeru commented on lib/CompilerImpl/ClangLibCompiler.cpp:
```diff
@@ -427,14 +433,14 @@ std::string ClangLibCompiler::runOptimizer(const std::string &src_IR,
   std::unique_ptr<llvm::ToolOutputFile> Out =
-    llvm::make_unique<llvm::ToolOutputFile>(optBCfilename,
+    std::make_unique<llvm::ToolOutputFile>(optBCfilename,
                                             outFileCreationErrorCode,
```
```suggestion
                                           outFileCreationErrorCode,
```

skeru commented on lib/CompilerImpl/ClangLibCompiler.cpp:
```diff
@@ -427,14 +433,14 @@ std::string ClangLibCompiler::runOptimizer(const std::string &src_IR,
+    std::make_unique<llvm::ToolOutputFile>(optBCfilename,
                                             outFileCreationErrorCode,
-                                            llvm::sys::fs::F_None);
+                                            llvm::sys::fs::OF_None);
```
```suggestion
                                           llvm::sys::fs::OF_None);
```

skeru commented on CMakeLists.txt:
```diff
@@ -107,10 +108,20 @@ if (NOT CMAKE_BUILD_TYPE)
+set ( CMAKE_CXX_VISIBILITY_INLINES_HIDDEN ON )
+# set ( CMAKE_INTERPROCEDURAL_OPTIMIZATION ON) 
+# CMAKE_INTERPROCEDURAL_OPTIMIZATION applies different optimizations depending on the compiler. It breaks cross compiler compatibility (e.g. g++ with clang)
+# Disabling link time optimization.
```
```suggestion

# https://gcc.gnu.org/wiki/Visibility#line-14 - Much lower chance of symbol collision.
set ( CMAKE_CXX_VISIBILITY_INLINES_HIDDEN ON )

# CMAKE_INTERPROCEDURAL_OPTIMIZATION applies different optimizations depending on the compiler. It breaks cross compiler compatibility (e.g. g++ with clang)
# Disabling link time optimization.
# set ( CMAKE_INTERPROCEDURAL_OPTIMIZATION ON) 
```

Align comments to the line below

skeru commented on CMakeLists.txt:
```diff
@@ -107,10 +108,20 @@ if (NOT CMAKE_BUILD_TYPE)
+set ( CMAKE_POSITION_INDEPENDENT_CODE ON )
+# https://gcc.gnu.org/wiki/Visibility#line-14 - Much lower chance of symbol collision.
+set ( CMAKE_CXX_VISIBILITY_INLINES_HIDDEN ON )
+# set ( CMAKE_INTERPROCEDURAL_OPTIMIZATION ON) 
+# CMAKE_INTERPROCEDURAL_OPTIMIZATION applies different optimizations depending on the compiler. It breaks cross compiler compatibility (e.g. g++ with clang)
```
That requires a bit further investigation/explanation. Which target breaks compatibility with which other?

Focshole replied:
https://github.com/tdlib/td/issues/1244#issuecomment-727657360
LTO is platform and compiler dependent, if the library is being used with another compiler breaks most stuff (gnu compilers and clang clash with weird errors such as :
```
/usr/bin/ld: /usr/local/lib/libVersioningCompiler.a: error adding symbols: file format not recognized
collect2: error: ld returned 1 exit status```

skeru commented on CMakeLists.txt:
```diff
@@ -202,19 +216,58 @@ endif (ENABLE_CLANG_AS_LIB)
+
+  set (VC_EXE_NAME libVC_testJit)
+
+  if ( CMAKE_BUILD_TYPE MATCHES DEBUG )
+    set_source_files_properties(${VC_TESTJIT_APP_SRC} PROPERTIES
+              COMPILE_FLAGS "-DVC_DEBUG -DHAVE_CLANG_AS_LIB -DFORCED_PATH_TO_TEST=\\\"${CMAKE_CURRENT_SOURCE_DIR}/test_code\\\""
+  )
+  else( CMAKE_BUILD_TYPE MATCHES DEBUG )
+    set_source_files_properties(${VC_TESTJIT_APP_SRC} PROPERTIES
+              COMPILE_FLAGS "-DHAVE_CLANG_AS_LIB -DFORCED_PATH_TO_TEST=\\\"${CMAKE_CURRENT_SOURCE_DIR}/test_code\\\""
+    )
+  endif( CMAKE_BUILD_TYPE MATCHES DEBUG )
```
Can this be restructured to add a single flag for each set method? I think some can be grouped outside this `if`

skeru commented on CMakeLists.txt:
```diff
@@ -107,10 +108,22 @@ if (NOT CMAKE_BUILD_TYPE)
+# Disabling link time optimization.
+# set ( CMAKE_INTERPROCEDURAL_OPTIMIZATION ON) 
+set ( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-use-cxa-atexit -fno-rtti" )
+if (CMAKE_COMPILER_IS_GNUC OR CMAKE_COMPILER_IS_GNUCXX)
+ set ( CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3" )
+endif(CMAKE_COMPILER_IS_GNUC OR CMAKE_COMPILER_IS_GNUCXX)
```
The code is quite self-explanatory, yet the reason behind this is not.
Can you add a line explaining it?

Also, if this is only for `-O3`, possibly we can set compiler properties instead of touching cmake flags

Focshole replied:
My fault, -O3 is ok.

https://gitlab.kitware.com/cmake/cmake/-/issues/23374
There is still no cmake config as of today

skeru commented on CMakeLists.txt:
```diff
@@ -107,10 +108,22 @@ if (NOT CMAKE_BUILD_TYPE)
+# CMAKE_INTERPROCEDURAL_OPTIMIZATION applies different optimizations depending on the compiler. It breaks cross compiler compatibility (e.g. g++ with clang)
+# Disabling link time optimization.
+# set ( CMAKE_INTERPROCEDURAL_OPTIMIZATION ON) 
+set ( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-use-cxa-atexit -fno-rtti" )
```
can these be moved to cmake settings instead of flags?

Focshole replied:
No, I'll add a comment explaining why (there is no cmake setting for those, one of those is an open issue)

skeru commented on CMakeLists.txt:
```diff
@@ -107,10 +108,22 @@ if (NOT CMAKE_BUILD_TYPE)
+set ( CMAKE_CXX_STANDARD 14 )
+set ( CMAKE_CXX_STANDARD_REQUIRED ON )
+set ( CMAKE_CXX_EXTENSIONS OFF )
+set ( CMAKE_VISIBILITY_INLINES_HIDDEN ON )
+set ( CMAKE_POSITION_INDEPENDENT_CODE ON )
+
+# https://gcc.gnu.org/wiki/Visibility#line-14 - Much lower chance of symbol collision.
+set ( CMAKE_CXX_VISIBILITY_INLINES_HIDDEN ON )
```
do we need two settings for visibility hidden or one depends on the other?

Focshole replied:
Nope, my fault

skeru commented on CMakeLists.txt:
```diff
@@ -202,19 +218,58 @@ endif (ENABLE_CLANG_AS_LIB)
-                              )
+  if ( CMAKE_BUILD_TYPE MATCHES DEBUG )
+    set_source_files_properties(${VC_TEST_APP_SRC} PROPERTIES
+                                COMPILE_FLAGS "-DVC_DEBUG -DHAVE_CLANG_AS_LIB -DFORCED_PATH_TO_TEST=\\\"${CMAKE_CURRENT_SOURCE_DIR}/test_code\\\""
+                                )
+  else ( CMAKE_BUILD_TYPE MATCHES DEBUG )
+    set_source_files_properties(${VC_TEST_APP_SRC} PROPERTIES
+                                COMPILE_FLAGS "-DHAVE_CLANG_AS_LIB -DFORCED_PATH_TO_TEST=\\\"${CMAKE_CURRENT_SOURCE_DIR}/test_code\\\""
+                                )
+  endif ( CMAKE_BUILD_TYPE MATCHES DEBUG )
 else (ENABLE_CLANG_AS_LIB)
-  set_source_files_properties(${VC_TEST_APP_SRC} PROPERTIES
-                              COMPILE_FLAGS "-flto -DFORCED_PATH_TO_TEST=\\\"${CMAKE_CURRENT_SOURCE_DIR}/test_code\\\""
-                              )
+  if ( CMAKE_BUILD_TYPE MATCHES DEBUG )
+    set_source_files_properties(${VC_TEST_APP_SRC} PROPERTIES
+                                COMPILE_FLAGS "-DVC_DEBUG -DFORCED_PATH_TO_TEST=\\\"${CMAKE_CURRENT_SOURCE_DIR}/test_code\\\""
+                                )
+  else ( CMAKE_BUILD_TYPE MATCHES DEBUG )
+    set_source_files_properties(${VC_TEST_APP_SRC} PROPERTIES
+                                COMPILE_FLAGS "-DFORCED_PATH_TO_TEST=\\\"${CMAKE_CURRENT_SOURCE_DIR}/test_code\\\""
+                                )
+  endif ( CMAKE_BUILD_TYPE MATCHES DEBUG )
```
possibly this can be rewritten with `target_compile_definitions` or `COMPILE_DEFINITIONS` property.

skeru commented on CMakeLists.txt:
```diff
@@ -231,13 +286,25 @@ set ( VC_EXE2_NAME libVC_testUtils )
-                              )
+  if ( CMAKE_BUILD_TYPE MATCHES DEBUG )
+    set_source_files_properties(${VC_TEST2_APP_SRC} PROPERTIES
+                                COMPILE_FLAGS "-DVC_DEBUG -DHAVE_CLANG_AS_LIB -DFORCED_PATH_TO_TEST=\\\"${CMAKE_CURRENT_SOURCE_DIR}/test_code\\\""
+                                )
+  else ( CMAKE_BUILD_TYPE MATCHES DEBUG )
+    set_source_files_properties(${VC_TEST2_APP_SRC} PROPERTIES
+                                COMPILE_FLAGS "-DHAVE_CLANG_AS_LIB -DFORCED_PATH_TO_TEST=\\\"${CMAKE_CURRENT_SOURCE_DIR}/test_code\\\""
+                                )
+  endif ( CMAKE_BUILD_TYPE MATCHES DEBUG )
 else (ENABLE_CLANG_AS_LIB)
-  set_source_files_properties(${VC_TEST2_APP_SRC} PROPERTIES
-                              COMPILE_FLAGS "-flto -DFORCED_PATH_TO_TEST=\\\"${CMAKE_CURRENT_SOURCE_DIR}/test_code\\\""
-                              )
+  if ( CMAKE_BUILD_TYPE MATCHES DEBUG )
+    set_source_files_properties(${VC_TEST2_APP_SRC} PROPERTIES
+                                COMPILE_FLAGS "-DVC_DEBUG -DFORCED_PATH_TO_TEST=\\\"${CMAKE_CURRENT_SOURCE_DIR}/test_code\\\""
+                                )
+  else ( CMAKE_BUILD_TYPE MATCHES DEBUG )
+    set_source_files_properties(${VC_TEST2_APP_SRC} PROPERTIES
+                                COMPILE_FLAGS "-DFORCED_PATH_TO_TEST=\\\"${CMAKE_CURRENT_SOURCE_DIR}/test_code\\\""
+                                )
+  endif ( CMAKE_BUILD_TYPE MATCHES DEBUG )
 endif (ENABLE_CLANG_AS_LIB)
```
Same as above about `target_compile_definitions`

skeru commented on Test.cpp:
```diff
@@ -91,7 +91,7 @@ int main(int argc, char const *argv[])
                                           "/usr/bin",
-                                          "/usr/lib/llvm/bin"
+                                          "/usr/bin"
                                         );
```
These initialisations are traditionally a pain.
Is there a way to get some default values being set via cmake?
(spoiler: yes)
Is there a way to get some default values being set via cmake without writing 3000 lines of cmake?
To be investigated....

Focshole replied:
Yes and yes, I'll push that soon

Focshole commented:
I am refactoring this project's cmake files. They work on my setup but I do expect it to fail on other systems. It is a wip right now.
Issues so far:
 - It fails to invoke the proper FindLLVM.cmake which should be shipped with every LLVM installation. It invokes an older version. I'll work on that.

Focshole commented:
It is working on ubuntu focal, I am looking for a way to run those tests on Travis CI's focal too, https://app.travis-ci.com/github/Focshole/libVersioningCompiler/builds/257609084 .
It looks like if it during those tests it works on a bad/poorly configured clang version. I'll have to investigate that.

Focshole commented on Test.cpp:
```diff
@@ -86,12 +98,12 @@ int main(int argc, char const *argv[])
+                                          LLVM_TOOLS_BINARY_DIR
                                         );
 #if HAVE_CLANG_AS_LIB
   vc::compiler_ptr_t clangAsLib = vc::make_compiler<vc::ClangLibCompiler>(
```
This is the part which probably breaks stuff in Travis CI. I have no clue on what could be wrong here. 
Reference:
https://app.travis-ci.com/github/Focshole/libVersioningCompiler/jobs/588014903#L622

skeru replied:
problem is the clang Driver that does not correctly passes options to the clang compiler. See `ClangLibCompiler::generateIR` around `ClangLibCompiler.cpp:118-151`

skeru replied:
Does this error happen only in the CI or also locally?
CI might be missing libClang or something like that

skeru commented on CMakeLists.txt:
```diff
@@ -107,10 +110,20 @@ if (NOT CMAKE_BUILD_TYPE)
+set ( CMAKE_CXX_STANDARD 14 )
+set ( CMAKE_CXX_STANDARD_REQUIRED ON )
+set ( CMAKE_CXX_EXTENSIONS OFF )
+# https://gcc.gnu.org/wiki/Visibility#line-14 - Much lower chance of symbol collision.
+set ( CMAKE_VISIBILITY_INLINES_HIDDEN ON )
+set ( CMAKE_POSITION_INDEPENDENT_CODE ON )
+# CMAKE_INTERPROCEDURAL_OPTIMIZATION applies different optimizations depending on the compiler.
+# It breaks cross compiler compatibility (e.g. library compiled with g++ which is being used with clang later on), so
+# I'm not enabling link time optimization by default.
+# set ( CMAKE_INTERPROCEDURAL_OPTIMIZATION ON) 
```
```suggestion

# https://gcc.gnu.org/wiki/Visibility#line-14 - Much lower chance of symbol collision.
set ( CMAKE_VISIBILITY_INLINES_HIDDEN ON )
set ( CMAKE_POSITION_INDEPENDENT_CODE ON )

# CMAKE_INTERPROCEDURAL_OPTIMIZATION applies different optimizations depending on the compiler.
# It breaks cross compiler compatibility (e.g. library compiled with g++ which is being used with clang later on), so
# I'm not enabling link time optimization by default.
# set ( CMAKE_INTERPROCEDURAL_OPTIMIZATION ON) 
```

skeru commented on CMakeLists.txt:
```diff
@@ -202,51 +220,72 @@ endif (ENABLE_CLANG_AS_LIB)
+target_compile_definitions(${VC_EXE_NAME} PRIVATE -DFORCED_PATH_TO_TEST="${CMAKE_CURRENT_SOURCE_DIR}/test_code")
+target_compile_definitions(${VC_EXE_NAME} PRIVATE -DLLVM_TOOLS_BINARY_DIR="${LLVM_TOOLS_BINARY_DIR}")
+
+if (ENABLE_CLANG_AS_LIB)
+  target_compile_definitions(${VC_EXE_NAME} PRIVATE -DHAVE_CLANG_AS_LIB)
+endif (ENABLE_CLANG_AS_LIB)
+if ( CMAKE_BUILD_TYPE MATCHES DEBUG )
+  target_compile_definitions(${VC_EXE_NAME} PRIVATE -DVC_DEBUG)
+endif ( CMAKE_BUILD_TYPE MATCHES DEBUG )
```
Can you please uniform across the file the formatting?
Having / not having a blank space between the `(` and the condition should be uniformed. No preference on which style to pick.

skeru commented on include/versioningCompiler/JITUtils.hpp:
```diff
@@ -0,0 +1,101 @@
+    compiler_ptr_t _libVC_jit_compiler;
+
+    llvm::TargetMachine *_targetMachine;
+/** Default compiler to be used in this helper. *
+//    compiler_ptr_t libVC_jit_compiler;
+
```
unclear

Focshole replied:
i think it was some forgotten comment

skeru commented on lib/CompilerImpl/ClangLibCompiler.cpp:
```diff
@@ -282,32 +285,35 @@ std::string ClangLibCompiler::runOptimizer(const std::string &src_IR,
-                                                 getRelocModel(),
-                                                 getCodeModel(),
+                                                 llvm::codegen::getRelocModel(),
+                                                 llvm::CodeModel::Small,// llvm::codegen::getCodeModel returns zero (casted to Tiny), which is wrong! Going with small which is the default model for majority of supported targets
```
does this happen on all versions of llvm?

Focshole replied:
I'll test it and I will update you about it :+1:

Focshole replied:
I get the commented behaviour on llvm 13,14 and 15.
`LLVM ERROR: Target does not support the tiny CodeModel`

skeru replied:
Can you try with `getExplicitCodeModel()`, which returns an optional, and fallback to Small only if not available?

skeru commented on lib/CompilerImpl/ClangLibCompiler.cpp:
```diff
@@ -282,32 +285,35 @@ std::string ClangLibCompiler::runOptimizer(const std::string &src_IR,
                                                                lookupError);
   // Some modules don't specify a triple, and this is okay.
   if (!TheTarget) {
+
```
```suggestion
```

Focshole commented:
I have formatted the cmake files using [cmake-format](https://pypi.org/project/cmake-format) as `cmake-format --literal-comment-pattern '.+' --first-comment-is-literal -i filename.cmake` to keep a uniform syntax convention.

skeru commented on lib/CompilerImpl/ClangLLVM/FileLogDiagnosticConsumer.cpp:
```diff
@@ -190,8 +190,13 @@ void FileLogDiagnosticConsumer::HandleDiagnostic(DiagnosticsEngine::Level Level,
+      #else
+      if (FE)
+        MainFilename = std::string(FE->getName());
+      #endif
```
can we move the `ifdef` only to the condition?

skeru commented on include/versioningCompiler/Compiler.hpp:
```diff
@@ -97,25 +97,26 @@ class Compiler
 
-/** \brief Closes the binary shared object, set *handler to nullptr, and
+    virtual 
+    /** \brief Closes the binary shared object, set *handler to nullptr, and
```
can you please move the virtual after the documentation?

skeru commented on include/versioningCompiler/CompilerImpl/JITCompiler.hpp:
```suggestion
```

skeru commented on include/versioningCompiler/CompilerImpl/JITCompiler.hpp:
should I stay or should I go?

Focshole replied:
This should be useful to add cross compile support. So far I haven't been able to integrate it seamingless, due to how llvm orc is made and my lack of knowledge in how does it work in its entirety. Should I focus on adding cross compilation features?

skeru replied:
I don't like to keep a constructor parameter commented. Either in or out.

If it is useful, we could possibly keep it as a default-initialised parameter, with default value being `native`-ish.

That being said, I do not expect this library to support cross-compilation out-of-the-box in the JIT compiler.
JIT kind of assumes you are already running on the target architecture, and there set of use cases featuring JIT-cross-compilation is ... limited to be generous.

Focshole replied:
I'd go with out.
[Here's why](https://github.com/skeru/libVersioningCompiler/pull/3/files#diff-b30d4a80eeb2d09e85cbb5adfa858ca07d880aca7628c5072860fa618d7352f9R80).

That parameter cannot be kept as such due to how modern orc works. To have a JITTargetMachineBuilder, you need to get a TargetTriple (getTargetTriple), which is being supplied (in this implementation ) by an ExecutorProcessControl, which is contained in an ExecutionSession, which should be generated (I guess) from a SelfExecutorProcessControl (the main process). At least right now this is how this is being done, I don't know how to simplify that section of code honestly

skeru replied:
ok, please remove

skeru commented on Test.cpp:
```diff
@@ -47,6 +47,18 @@
 #define TEST_FUNCTION_LBL "TEST_FUNCTION"
 #endif
 
+#ifndef LLVM_TOOLS_BINARY_DIR
+#define LLVM_TOOLS_BINARY_DIR "/usr/bin"
+#endif
+
+#ifndef CLANG_EXE_NAME
+#define CLANG_EXE_NAME "clang"
+#endif
+
+#ifndef OPT_EXE_FULLPATH
+#define OPT_EXE_FULLPATH "/usr/bin/opt"
+#endif
+
```
I think these define (and previous ones above) might be better suited to become `const std::string`s initialised with the define values.

skeru replied:
That might help (if possible) to start adding some code to verify the existence of such files via `std::filesystem:path` in the future, should we ever move to require C++17

Focshole replied:
Those variables are distribution/machine dependent due to where are actually opt and clang located (taking into account of many possible installed versions, which is by default what happens in Ubuntu for example). The cmake file provides those paths, the ifndef should always be false (all of those variables should always be defined).

Those paths seemed to be used a lot at runtime, that's why I preferred to move those checks in cmake/*.cmake.

I'll look to replace most path composition functions with standard ones :+1:

skeru replied:
I agree on having the checks in cmake. I was mostly arguing in favour of having define-initialised constants rather than simple defines in the code.

Focshole replied:
To be honest, I have add them to make intellisense happy (which otherwise couldn't find the variable definition) :smile:

skeru commented on CMakeLists.txt:
```diff
@@ -18,30 +18,35 @@
 # along with libVersioningCompiler. If not, see <http://www.gnu.org/licenses/>
 
-cmake_minimum_required(VERSION 3.0.2)
+cmake_minimum_required(VERSION 3.20)
```
probably worth having a comment with the latest feature / behaviour we are using to require this version.

skeru commented on CMakeLists.txt:
```diff
@@ -18,30 +18,35 @@
-project(libVersioningCompiler)
+project(
+  libVersioningCompiler
+  VERSION 2.0
```
```suggestion
  VERSION 3.0
```
Have a look at `git tag`. There already is a `v2.0`
I propose to name it `3.0`.

skeru commented on CMakeLists.txt:
```diff
@@ -50,203 +55,236 @@ link_directories(${UUID_LIBRARY})
+if(ENABLE_CLANG_AS_LIB)
+  set_source_files_properties(
+    ${VC_LIB_HDR_PREFIX}/Features.hpp PROPERTIES COMPILE_FLAGS
+                                                 "-DENABLE_CLANG_AS_LIB=true")
```
can we use [`COMPILE_DEFINITIONS`](https://cmake.org/cmake/help/latest/prop_tgt/COMPILE_DEFINITIONS.html) instead?

skeru commented on README.md:
```diff
@@ -11,15 +12,15 @@ An introductory video is available on [YouTube](https://www.youtube.com/watch?v=
- - cmake 3.0.2 or greater
+ - Ubuntu 20.04 LTS or greater
+ - any compiler compliant to the C++14 standard
+ - cmake 3.9.4 or greater
```
not coherent with CMakeLists.txt

skeru commented on README.md:
```diff
@@ -11,15 +12,15 @@ An introductory video is available on [YouTube](https://www.youtube.com/watch?v=
- - (OPTIONAL) llvm-8.0-dev
- - (OPTIONAL) libclang-8.0-dev
+ - (OPTIONAL) LLVM 13 or greater (up to LLVM 15)
+ - (OPTIONAL) libclang-13-dev up to libclang-15-dev
```
```suggestion
 - (OPTIONAL) LLVM 13 or greater (tested up to LLVM 15)
 - (OPTIONAL) libclang-13-dev or greater (tested up to libclang-15-dev)
```

skeru commented on README.md:
```diff
@@ -47,7 +59,7 @@ You will find a pre-cooked cmake module in `${LIBVC_ROOT}/config`.
+$LIBVC_INCLUDE_DIRS   = Include path for the header files of libVersioningCompiler
 $LIBVC_LIBRARIES  = Link these libraries to use libVersioningCompiler
 $LIVC_LIB_DIR     = Extra libraries directories
 $HAVE_CLANG_LIB_COMPILER = Set to true if libVersioningCompiler can have
```
can we align the variable descriptions to the rightmost `=` symbol?

skeru commented on include/versioningCompiler/CompilerImpl/ClangLLVM/LLVMInstanceManager.hpp:
```diff
@@ -30,11 +33,23 @@
+#include "llvm/Support/Program.h"
 
 #include <string>
+#include <filesystem>
```
```suggestion
#include <filesystem>
#include <string>
```

skeru commented on include/versioningCompiler/CompilerImpl/ClangLLVM/LLVMInstanceManager.hpp:
```diff
@@ -30,11 +33,23 @@
+#include "llvm/InitializePasses.h"
+#include "llvm/Support/ManagedStatic.h"
+#include "llvm/ADT/APInt.h"
+#include "llvm/Support/Program.h"
```
please sort

skeru commented on include/versioningCompiler/CompilerImpl/SystemCompilerOptimizer.hpp:
```diff
@@ -35,25 +35,25 @@ class SystemCompilerOptimizer : public SystemCompiler
+                          const std::filesystem::path libWorkingDir,
+                          const std::filesystem::path log = "",
+                          const std::filesystem::path installDir = "/usr/bin",
+                          const std::filesystem::path optimizerInstallDir = "/usr/bin");
```
these initialisations are not coherent with those of SystemCompiler.hpp, aka using `std::filesystem::u8path()`.
Please consider having the same constructor here, or removing it from there.

skeru commented on include/versioningCompiler/Option.hpp:
```diff
@@ -39,7 +39,7 @@ class Option
    * This string should include also separators such as " ", "=".
    * Examples of optionPrefix are:
-   * "-O" for `-O2`, "--std=" for `--std=c++11`, "--emit-llvm" for `--emit-llvm`
+   * "-O" for `-O2`, "--std=" for `--std=c++14`, "--emit-llvm" for `--emit-llvm`
```
does this actually make a difference?

skeru commented on include/versioningCompiler/Version.hpp:
```diff
@@ -30,6 +30,7 @@
 #include <unordered_map>
 #include <vector>
 #include <uuid/uuid.h>
+#include <filesystem>
```
please sort

skeru commented on lib/Compiler.cpp:
```diff
@@ -23,17 +23,18 @@
 #include <dlfcn.h> // needed for loadSymbol
 #include <fstream>
 #include <sys/stat.h>
+#include <filesystem>
```
please sort

Focshole replied:
I should have sorted all includes, let me know if there is something you don't like!

skeru commented on lib/Compiler.cpp:
```diff
@@ -165,63 +166,53 @@ bool Compiler::exists(const std::string &name)
-  filename = filename + "IR_";
-  filename = filename + versionID;
-  filename = filename + ".bc";
+  std::filesystem::path filename = libWorkingDirectory / std::filesystem::u8path("IR_"+ versionID+ ".bc");
```
This `operator /` has a tricky [behaviour](https://en.cppreference.com/w/cpp/filesystem/path/append). Can you please add a couple of test cases to verify it is working correctly both with relative and with absolute `libWorkingDirectory` paths?

Focshole replied:
I have added some text warnings and I have checked that they all work properly. Let me know what you think about it!

skeru commented on lib/CompilerImpl/ClangLibCompiler.cpp:
```diff
@@ -463,7 +471,7 @@ std::string ClangLibCompiler::runOptimizer(const std::string &src_IR,
     Passes.add(createPrintModulePass(*OS, "", PreserveAssemblyUseListOrder));
   } else if (OutputThinLTOBC) {
-    Passes.add(createWriteThinLTOBitcodePass(*OS));
+    Passes.add(createWriteThinLTOBitcodePass(*OS)); // TODO not sure if this is what causes the optimization errors
```
just a reminder to have a look at this TODO

Focshole replied:
It wasn't an issue in the end, I'll remove the comment

skeru commented on Test.cpp:
```diff
@@ -86,12 +98,12 @@ int main(int argc, char const *argv[])
-                                          "./test.log",
-                                          "/usr/bin",
-                                          "/usr/lib/llvm/bin"
+                                          "./test.log", 
```
can this one go to `filesystem` style as well?

Focshole replied:
If you prefer, I'll do it but it won't change much, it will still invoke the std::filesystem constructor and it won't need the std::filesystem::u8path (which is deprecated in c++20 but required in c++17) due to the fact that it does not contain "risky" utf chars

skeru replied:
What I would like to achieve is a consistent format across all the project files.

If the C++ standard creates an issue, we can have an helper header containing something like

```
#if defined(__cplusplus) && __cplusplus <= 201703L

using vc::mypath = std::filesystem::u8path;

#else

using vc::mypath = std::string;

#endif
```

Focshole replied:
Ok, I had replaced them all with std::filesystem::u8path!

skeru commented on Test.cpp:
```diff
@@ -103,6 +115,7 @@ int main(int argc, char const *argv[])
   // ---------- End compilers initialization ----------
   // start configuring version v
   builder._compiler = cc;
+  builder._genIROptionList={vc::Option("fpic", "-fPIC")};
```
```suggestion
  // avoid relocation problems with global variables
  builder._genIROptionList={vc::Option("fpic", "-fPIC")};
```

skeru reviewed (changes requested):
I tested the code and it seems to work.
Only minor changes required before merging

danielecattaneo commented on CMakeLists.txt:
```diff
@@ -18,30 +18,35 @@
-find_package(DL REQUIRED)
-include_directories(${DL_INCLUDE_DIR})
-link_directories(${DL_LIBRARY})
+# check for DL library
+if(NOT CMAKE_DL_LIBS)
+  message(FATAL_ERROR "Missing DL libs!")
+endif(NOT CMAKE_DL_LIBS)
```
This check always fails on macOS because the dynamic linker is always used and libdl.so is not used/needed. You need to explicitly disable the check on Apple platforms like this:
```
if(NOT ${APPLE} AND NOT CMAKE_DL_LIBS)
...
endif()
```

danielecattaneo commented on CMakeLists.txt:
```diff
@@ -50,203 +55,236 @@ link_directories(${UUID_LIBRARY})
-set (VC_LIB_DEPS ${DL_LIBRARY}
-                 ${UUID_LIBRARY}
-                )
+set(VC_LIB_DEPS ${CMAKE_DL_LIBS} ${UUID_LIBRARY} stdc++)
```
stdc++ should not be needed here, and it causes problems on platforms that use LLVM's own libc++ (which is linked as just -lc++ instead).

Focshole replied:
Well, it seems required to build the tests, without it I'd get linking errors. 
I had attempted a fix which adds stdc++ on linux and c++ elsewhere. Let me know if it seems a proper one!

danielecattaneo commented on cmake/FindLLVM_config.cmake:
```diff
@@ -0,0 +1,199 @@
+  get_filename_component(OPT_EXE_NAME ${OPT_EXE_FULLPATH} NAME)
+endif(NOT OPT_EXE_NAME)
+
+set(LLVM_SYSTEM pthread z tinfo)
```
This list of libraries depends on how LLVM is built and therefore cannot be hardcoded. Use `llvm-config --system-libs` to get the list instead of hard-coding it.

Focshole replied:
I didn't know where do those settings came out from and why they were needed, thank you for the explaination!
Fixed

danielecattaneo commented on cmake/FindLibClang.cmake:
```diff
@@ -1,152 +1,167 @@
+  endif(CLANG_${_libname_}_LIB)
+endmacro(FIND_AND_ADD_CLANG_LIB)
+
+set(LIBCLANG_LIBRARIES)
+# check if is being linked as shared or not
+if(NOT LLVM_CONFIG_EXECUTABLE)
+  find_program(
+    LLVM_CONFIG_EXECUTABLE
+    NAMES "llvm-config-${LLVM_VERSION_MAJOR}" "llvm-config"
+    DOC "llvm-config executable"
+    PATHS ${LLVM_TOOLS_BINARY_DIR})
+endif(NOT LLVM_CONFIG_EXECUTABLE)
```
The comment says we're checking if clang should be linked as shared or not, but no check is done. This check however is needed because if you build LLVM with LLVM_BUILD_LLVM_DYLIB=ON, then

- both static and dynamic libraries will be available for llvm and clang
- llvm-config will return the static libraries by default (at least on macOS)
- FindLibClang.cmake will find libclang-cpp.so which is dynamic

Since both libLLVM.so (through libclang) and the static LLVM libraries are linked together this causes double definitions and in general the built binaries to fail.

You need to run `llvm-config --shared-mode` and check what it returns before searching for dynamic/static clang libraries.

Focshole replied:
I tried this approach, I am testing it with dynamic builds ("shared" ones) but I am failing in being able to test it with a static build. I had found a static build ready [here](https://github.com/savi-lang/llvm-static).

I do either get linking errors when using my compiler such as
```
libclangFrontend.a(TextDiagnostic.cpp.o): relocation R_X86_64_32S against symbol `_ZTVN5clang14TextDiagnosticE' can not be used when making a PIE object; recompile with -fPIE
``` 
Or I get compiling errors such as:
```
/usr/include/wchar.h:35:10: fatal error: 'stddef.h' file not found
```
when using the provided compiler as CXX.

I'll try to make a static build myself of llvm, it is going to take quite some time I guess

Focshole replied:
Ok, I had been able to test libVersioningCompiler on MacOS. There is a new section on the readme about how to build it on MacOS. Let me know if it works as expected!

Focshole replied:
> The libLLVM linking mode should also apply to libclang.

https://github.com/Focshole/libVersioningCompiler/blob/c4c932bed7dd0f156adee5cd80daefe7d33e65f1/cmake/FindLibClang.cmake#L88

I guess this doesn't work.  Do you have any logs with the verbose flags? Which library did cmake chosen to link against?

danielecattaneo reviewed (commented):
Some comments on bugs I found after @skeru asked me to try to build libVC on macOS.

skeru commented on README.md:
````diff
@@ -28,16 +29,27 @@ like the Clang-as-a-library compiler implementation.
 
 libVersioningCompiler uses the cmake build system.
 If you want to install libVersioningCompiler
-
+install the dependencies with LLVM version `${LLVM_V}`:
+
+```bash
+wget https://apt.llvm.org/llvm.sh
+chmod +x llvm.sh
+[sudo] ./llvm.sh ${LLVM_V} all
+[sudo] apt-get install -y git cmake uuid-dev g++-10 libz-dev
+[optional][sudo] update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-${LLVM_V} 100
+[optional][sudo] update-alternatives --install /usr/bin/clang clang /usr/bin/clang-${LLVM_V} 100
+[optional][sudo] update-alternatives --install /usr/bin/opt opt /usr/bin/opt-${LLVM_V} 100
+```
````
Need to add a section

```
OR manually specify your path to pre-installed LLVM version using
export SOME_LLVM_DIRECTORY=/path/to/my/installation
```

Focshole replied:
I had updated the readme with how to specify a custom LLVM installation, let me know what you think about it

skeru commented on include/versioningCompiler/Compiler.hpp:
```diff
@@ -94,28 +95,29 @@ class Compiler
-                                 void ** handler) const;
+                                 void ** handler);
 
+     
```
```suggestion
```

skeru commented on include/versioningCompiler/CompilerImpl/ClangLLVM/LLVMInstanceManager.hpp:
```diff
@@ -21,19 +24,30 @@
+#include "llvm/IR/PassManager.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/PassRegistry.h"
+#include "llvm/Support/Host.h"
+#include "llvm/Support/ManagedStatic.h"
 #include "llvm/Support/MemoryBuffer.h"
 #include "llvm/Support/Path.h"
 #include "llvm/Support/Program.h"
-#include "llvm/Support/raw_os_ostream.h"
-#include "llvm/Support/TargetRegistry.h"
 #include "llvm/Support/TargetSelect.h"
+#include "llvm/Support/raw_os_ostream.h"
```
swap?

skeru commented on lib/Compiler.cpp:
```diff
@@ -165,63 +167,53 @@ bool Compiler::exists(const std::string &name)
 		const std::string &versionID,  
 		int incremental) const
-{
+{ 
```
```suggestion
{
```

Focshole commented:
I have fixed all formatting issues with clang-format, tested the code on MacOS and added notes on how to do it. Let me know if am I missing something!

Focshole commented:
The Jit does not work as expected. When doing a Debug build (on llvm 15.0.7), it crashes with:
```
IR prepared. Going for myversion compilation..
Failure value returned from cantFail wrapped call
Duplicate definition of symbol 'global_var'
UNREACHABLE executed at /usr/include/llvm/Support/Error.h:755!
zsh: IOT instruction (core dumped)  ./libVersioningCompiler/build/libVC_testJit
```

When running as a release, it won't crash but valgrind detects a memory leak due to that specific error message which is not being handled properly in OrcV2 (because it gets suppressed in release).

[This](https://discourse.llvm.org/t/duplicate-symbols-in-separate-modules-no-longer-allowed/1805) should be the relevant issue but I have no clue on how to fix this.

Focshole commented:
I had fixed all memory leaks and crashes but I am not sure if that is the proper way to handle the duplicate definition issue.

danielecattaneo commented:
> I have fixed all formatting issues with clang-format, tested the code on MacOS and added notes on how to do it. Let me know if am I missing something!

Now it builds without errors. But there's another issue: on macOS llvm-config is buggy and does not return libLLVM even when it exists. Therefore the libVC custom configuration scripts can find LLVM's static modules and clang's dynamic module and link them together, causing double symbol definitions. I.e. when i run `./libVC_testJit` or `./libVC_test` I get the message

```
: CommandLine Error: Option 'verify-noalias-scope-decl-dom' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
Abort trap: 6
```

The libLLVM linking mode should also apply to libclang.

Focshole commented:
Functionally the code is ok to me, I'd improve the tests/interfaces on compilers and versions to be cleaner and clearer (many tests access _variables to set parameters).

Focshole commented:
> Functionally the code is ok to me, I'd improve the tests/interfaces on compilers and versions to be cleaner and clearer (many tests access _variables to set parameters).

Done.

Focshole commented:
I have refactored a bit the history, so that it would be much easier to find read edits

skeru commented on test_code/test_code.c:
```diff
@@ -22,16 +22,15 @@
 #ifdef TEST_FUNCTION
 
-float global_var;
+float global_var = -1;
```
```suggestion
float global_var = -1.0f;
```
